### PR TITLE
Clarify HTTP header value encoding and decoding requirements

### LINKED ISSUE #777: Clarify HTTP header encoding/decoding rules
The [HTTP protocol binding spec](https://github.com/cloudevents/spec/blob/v1.0.1/http-protocol-binding.md#3132-http-header-values) states:

> String values MUST be percent-encoded as described in [RFC3986, section 2.4](https://tools.ietf.org/html/rfc7230#section-3) before applying the header encoding rules described in
[RFC7230, section 3.2.6](https://tools.ietf.org/html/rfc7230#section-3.2.6).

I'm finding it confusing to implement this appropriately and rigorously in the C# SDK.

To start with, I find it a little odd that we're referring to a URI-specific RFC, when the values being encoded are arbitrary strings, not necessarily URIs. This immediately presents a problem when referring to RFC 3986 section 2.4, which starts: 

> Under normal circumstances, the only time when octets within a URI are percent-encoded is during the process of producing the URI from its
> component parts.  This is when an implementation determines which of the reserved characters are to be used as subcomponent delimiters
> and which can be safely used as data.  Once produced, a URI is always in its percent-encoded form.

If I have an attribute value of `"xyz:abc://:123//"` (which looks a bit URI-like, but may not have any relation to real URIs), do the colons and slashes need escaping in that? I think it would be helpful to specify the encoding and decoding requirements entirely separately from URI encoding, even if we still end up using a percent-encoding scheme (which we'd need to for consistency, of course).

Ideally, I'd like to see some conformance tests here - including for values containing double quotes and commas, which could make the second part of the encoding (the HTTP-specific part) interesting. Effectively a list of examples with:

- Original canonical string representation of the CE attribute
- Percent-encoded (ASCII-only) text
- On-the-wire header value (complete with quoting etc where necessary)

If such conformance tests already exist somewhere, please let me know :)

### PR DESCRIPTION
**Pull request by jskeet**
Fixes #777

Once this is agreed, we should also document conformance tests, but
it's unclear at the moment where they should live.

Signed-off-by: Jon Skeet <jonskeet@google.com>

### CODE REVIEW

**Comment by slinkydeveloper** [comment]
cc @slinkydeveloper

File: http-protocol-binding.md [hunk 1]
```diff
@@ -213,16 +213,72 @@ and per [RFC7230, section 3][rfc7230-section-3], HTTP headers MUST only use
...
+described in [RFC7230, section 3.2.6][rfc7230-section-3-2-6] MUST be
+decoded to an ASCII string, and then a **single round** of
+percent-decoding as described below. HTTP headers for CloudEvent
+attribute values do not support parenthetical comments, so the
+initial decoding only needs to handle double-quoted values, including
+backslash escapes. Header values produced via the percent-encoding
+described here will never include double-quoted values, but they MUST
+be supported for compatibility with older versions of this
+specification which did not require double-quote and space
+characters to be percent-encoded.
```
Comment (slinkydeveloper) [inline http-protocol-binding.md:232]: That's the sentece that (I think) will allow old and new sdks to communicate. I like it.

Just, can you specify that _they MUST be supported when receiving events_ or something like that? Just to make clear the first part of the sentence is about producing, the second is about consuming

**Review by slinkydeveloper — COMMENTED** [review]
Sounds good to me

File: http-protocol-binding.md (hunk 1 above)
Comment (jskeet) [reply http-protocol-binding.md:232]: Yup, happy to do so. (Will do that in a separate commit, hopefully some time in the next few hours.)

Comment (slinkydeveloper) [reply http-protocol-binding.md]: Perfect!

**Comment by dazuma** [comment]
I'll look at wording in more detail in the next few days, but for now the overall approach looks good to me.

File: http-protocol-binding.md [hunk 2]
```diff
@@ -213,16 +213,72 @@ and per [RFC7230, section 3][rfc7230-section-3], HTTP headers MUST only use
...
-2.4][rfc3986-section-2-4] before applying the header encoding rules described in
-[RFC7230, section 3.2.6][rfc7230-section-3-2-6].
-
-When decoding an HTTP message into a CloudEvent, these rules MUST be applied in
-reverse -- [RFC7230, section 3.2.6][rfc7230-section-3-2-6] decoding to an ASCII
-string, and then a **single round** of percent-decoding as described in
-[RFC3986, section 2.4][rfc3986-section-2-4] to produce a valid UTF-8 String.
-(Note that applying percent-decoding an incorrect number of times can result in
-message corruption or security issues.)
+When encoding a CloudEvent as an HTTP message, string values MUST be
```
Comment (duglin) [inline http-protocol-binding.md:216]: All strings or just HTTP headers?

Comment (jskeet) [reply http-protocol-binding.md:216]: Just headers - thanks, will fix.

File: http-protocol-binding.md [hunk 3]
```diff
@@ -213,16 +213,72 @@ and per [RFC7230, section 3][rfc7230-section-3], HTTP headers MUST only use
...
+[RFC3986, section 2.1][rfc3986-section-2-1] but is more specific
+about what needs encoding. The resulting string does not then need
+any further quoting, due to space and quote characters being encoded.
+
+When decoding an HTTP message into a CloudEvent, quoted strings as
+described in [RFC7230, section 3.2.6][rfc7230-section-3-2-6] MUST be
+decoded to an ASCII string, and then a **single round** of
+percent-decoding as described below. HTTP headers for CloudEvent
+attribute values do not support parenthetical comments, so the
+initial decoding only needs to handle double-quoted values, including
```
Comment (duglin) [inline http-protocol-binding.md]: it's not clear to me if the "including" here means we need to process backslash escapes or whether they're ignored. I'm guessing it means to NOT do them. 

Why are comment not allowed? Do we talk about that someplace else?

File: http-protocol-binding.md [hunk 4]
```diff
@@ -213,16 +213,75 @@ and per [RFC7230, section 3][rfc7230-section-3], HTTP headers MUST only use
...
+represented as HTTP header values MUST be percent-encoded as
+described below. This is compatible with [RFC3986, section
+2.1][rfc3986-section-2-1] but is more specific about what needs
+encoding. The resulting string does not then need any further
+quoting, due to space and quote characters being encoded.
+
+When decoding an HTTP message into a CloudEvent, any quoted string
+as described in [RFC7230, section 3.2.6][rfc7230-section-3-2-6]
+appearing as an HTTP header value MUST be decoded to an ASCII
+string, and then a **single round** of percent-decoding must be
```
Comment (duglin) [inline http-protocol-binding.md:226]: s/must/MUST/

File: http-protocol-binding.md (hunk 3 above)
Comment (jskeet) [reply http-protocol-binding.md]: First part: clarified that it means processing. (Basically that's part of "handling double-quoted values".)

See https://github.com/cloudevents/sdk-csharp/pull/117/files#diff-66bbdbd26048de2dc60bc1ae88c9626d4928326756b34dc187e90133b202af13 for some examples.

In terms of comments: I don't think we talk about that anywhere, but I don't think they ever *have* been allowed, and I'm asserting we don't want to introduce that capability. This is the problem with RFC 7230 section 3.2.6 - it expects the header field values to be defined elsewhere, and we basically haven't done that in the past.

File: http-protocol-binding.md (hunk 4 above)
Comment (jskeet) [reply http-protocol-binding.md]: Done. (And obviously I'll squash all of this down into a single commit before we merge, so please excuse the jovial temporary commit messages.)

File: http-protocol-binding.md [hunk 5]
```diff
@@ -213,16 +213,75 @@ and per [RFC7230, section 3][rfc7230-section-3], HTTP headers MUST only use
...
+- Space (U+0020)
+- Double-quote (U+0022)
+- Percent (U+0025)
+- Any characters outside the printable ASCII range of U+0021-U+007E
+  inclusive
+
+Attribute values are already constrained to prohibit characters in
+the range U+0000-U+001F inclusive and U+007F-U+009F inclusive;
+however for simplicity and to account for potential future changes,
+it is encouraged that any HTTP header encoding implementation treats
```
Comment (duglin) [inline http-protocol-binding.md:252]: s/encouraged/RECOMMENDED/

**Comment by duglin** [comment]
I think there might have been a mistake in a rebase - you now have a lot of other files that are in conflict

**Comment by jskeet** [comment]
Eek! Thanks, will fix the rebase now.

**Comment by jskeet** [comment]
(Rebase fixed - I'd used "git rebase -i master" out of habit.)

**Comment by duglin** [comment]
will you be submitting a PR for 'master' ?

**Comment by duglin** [comment]
@jskeet to be explicit, what would change w.r.t. encoding based on this PR? Can you give an example of a header that might now look different based on these rules?

**Comment by jskeet** [comment]
> will you be submitting a PR for 'master' ?

I wasn't expecting to. Do we have a description anywhere of how branches in this repository are used?

**Comment by duglin** [comment]
hmm my memory is failing me. I seem to recall that we agreed that minor typos can go into the existing "dot releases" w/o any concern. But that bigger changes will need a new "dot release". So, should this really be directed to "v1.0.1" or "master" and then a "v1.0.2" created?

**Comment by jskeet** [comment]
> @jskeet to be explicit, what would change w.r.t. encoding based on this PR? Can you give an example of a header that might now look different based on these rules?

A header using double-quotes might previously have been (arguably-incorrectly) emitted without any encoding, e.g. for an extension attribute `xyz` with string value `They said "Hello!"`:

```text
ce-xyz: They said "Hello!"
```

Or correctly using double-quoting and backslash escaping:

```text
ce-xyz: They said "\"Hello!\""
```

This would now be emitted as:

```text
ce-xyz: They said %22Hello%22!
```

Any SDK parsing the previous broken version would read the attribute as `They said Hello!` (no quotes) but parsing the previous working (double quoted and backslash escaped) version the attribute would be read correctly.

The main purpose of this PR is to be much clearer in expected behavior.

> hmm my memory is failing me. I seem to recall that we agreed that minor typos can go into the existing "dot releases" w/o any concern. But that bigger changes will need a new "dot release". So, should this really be directed to "v1.0.1" or "master" and then a "v1.0.2" created?

Maybe? Let's discuss on the call. I'm happy to rebase the PR when we know what we want to do.

**Comment by duglin** [comment]
found it!
```
- Additionally, the working group has decided that "patch" changes that are
  trivial in nature (e.g. typos) can be merged directly into the latest
  release without a version number change and without creating a new formal
  "patch" branch/release. Non-trivial "patch" changes will necessitate the
  creation of a new "patch" branch (e.g. `1.0.2-wip`) and release cycle that
  will eventually result in an official "patch" release (e.g. `1.0.2`)
```

**Comment by duglin** [comment]
thanks for the example. Given the new rules, how would `e-xyz: They said "Hello!"` be interpreted? I think the text implies it's an error, but must be supported. But I'm not sure it says what to do - I'm guessing it would result in `They said Hello!`

**Comment by jskeet** [comment]
> Given the new rules, how would `e-xyz: They said "Hello!"` be interpreted? I think the text implies it's an error, but must be supported. But I'm not sure it says what to do - I'm guessing it would result in `They said Hello!`

Yes, that's the expected result. It wouldn't be an error, just unnecessary double-quoting (which almost certainly implies it is the result of broken processing, but isn't actually invalid).

Given your comment about patch changes, would you prefer this to go onto the `master` branch? Would it make sense to have a bigger talk in the working group about versioning at some point, so we can document the process and expectations more clearly? (I've found some aspects in the minutes, but not the bit you're quoting... it's currently unclear whether we should expect to see `ce-specversion:1.0.1` at some point etc. I'd love to see all this documented. And I realize that probably puts me on the hook for getting the ball rolling.)

**Comment by duglin** [comment]
Can certainly have a chat on the call.  In the meantime, see the last bullet here:  https://github.com/cloudevents/spec/blob/master/community/GOVERNANCE.md#release-process-and-versioning

**Comment by jskeet** [comment]
Aha - thanks very much for that. Will try to read and absorb before the meeting, as it may cover everything anyway.

**Comment by jskeet** [comment]
Note: when this has been reviewed by @clemensv, I'll rebase and retarget it to the master branch.
(I didn't want to do that while Clemens *is* reviewing, as that could cause confusion.)

File: http-protocol-binding.md [hunk 6]
```diff
@@ -213,16 +213,75 @@ and per [RFC7230, section 3][rfc7230-section-3], HTTP headers MUST only use
...
-reverse -- [RFC7230, section 3.2.6][rfc7230-section-3-2-6] decoding to an ASCII
-string, and then a **single round** of percent-decoding as described in
-[RFC3986, section 2.4][rfc3986-section-2-4] to produce a valid UTF-8 String.
-(Note that applying percent-decoding an incorrect number of times can result in
-message corruption or security issues.)
+When encoding a CloudEvent as an HTTP message, string values
+represented as HTTP header values MUST be percent-encoded as
+described below. This is compatible with [RFC3986, section
+2.1][rfc3986-section-2-1] but is more specific about what needs
+encoding. The resulting string does not then need any further
```
Comment (clemensv) [inline http-protocol-binding.md:220]: "does not then need" smells like a "MUST NOT"?

File: http-protocol-binding.md [hunk 7]
```diff
@@ -213,16 +213,75 @@ and per [RFC7230, section 3][rfc7230-section-3], HTTP headers MUST only use
...
+quoting, due to space and quote characters being encoded.
+
+When decoding an HTTP message into a CloudEvent, any quoted string
+as described in [RFC7230, section 3.2.6][rfc7230-section-3-2-6]
+appearing as an HTTP header value MUST be decoded to an ASCII
```
Comment (clemensv) [inline http-protocol-binding.md:225]: The quoted string is already US-ASCII per definition and with the kinda-sorta exceptions discussed in RFC7230, 3.2.4 in the last paragraph. I would just omit "decoded to an ASCII string" here because that's what it ought to be already and it's confusing what the decoding action ought to be.

"...appearing as an HTTP header value (limited to US-ASCII per RTC730) must be percent-decoded as described below" 

maybe?

File: http-protocol-binding.md (hunk 6 above)
Comment (jskeet) [reply http-protocol-binding.md:220]: Not particularly... it would be *okay* to quote things (it wouldn't break anything) but we wouldn't encourage SDKs to do that, as there's no point. (Just like it's *valid* for an SDK to escape every character, even ASCII alphanumeric ones. It's just pointless and wasteful.)

I'd be happy to add a "SHOULD NOT" for this, to make the discouragement clear?

File: http-protocol-binding.md [hunk 8]
```diff
@@ -213,16 +213,75 @@ and per [RFC7230, section 3][rfc7230-section-3], HTTP headers MUST only use
...
+string, and then a **single round** of percent-decoding MUST be
+performed as described below. HTTP headers for CloudEvent attribute
+values do not support parenthetical comments, so the initial
```
Comment (clemensv) [inline http-protocol-binding.md:228]: "parenthetical comments" ?

File: http-protocol-binding.md [hunk 9]
```diff
@@ -213,16 +213,75 @@ and per [RFC7230, section 3][rfc7230-section-3], HTTP headers MUST only use
...
+processing backslash escapes within double-quoted values. Header
+values produced via the percent-encoding described here will never
+include double-quoted values, but they MUST be supported when
+receiving events, for compatibility with older versions of this
+specification which did not require double-quote and space
+characters to be percent-encoded.
+
+Percent encoding is performed by considering each Unicode character
+within the attribute's canonical string representation. Any
+character represented in memory as a surrogate pair MUST be treated
```
Comment (clemensv) [inline http-protocol-binding.md:239]: "surrogate pair" requires a reference link

File: http-protocol-binding.md (hunk 7 above)
Comment (jskeet) [reply http-protocol-binding.md:225]: Right. I think this was a matter of keeping too much of what was already there.

But we do need to specify that "unquoting" is performed. How about this:

> When decoding an HTTP message into a CloudEvent, any HTTP header
> value MUST have quoted string escaping as described in
> [RFC7230, section 3.2.6][rfc7230-section-3-2-6] applied in reverse, and
> then a **single round** of percent-decoding MUST be performed as
> described below.

(Remainder of paragraph as-is.)

File: http-protocol-binding.md (hunk 8 above)
Comment (jskeet) [reply http-protocol-binding.md:228]: Yes, apparently some headers can include comments, which appear in parentheses. For example, here's a user agent header value:

```text
Mozilla/5.0 (Windows NT 6.1; Win64; x64; rv:47.0) Gecko/20100101 Firefox/47.0
```

In that, the `(Windows NT 6.1; Win64; x64; rv:47.0)` part is a comment. Only a few HTTP header fields are documented to support comments, and I suggest we never include CloudEvent headers in that set :)

File: http-protocol-binding.md (hunk 9 above)
Comment (jskeet) [reply http-protocol-binding.md:239]: Fine by me - is http://unicode.org/glossary/#surrogate_pair okay with you? That feels better than linking into a PDF.

File: http-protocol-binding.md (hunk 6 above)
Comment (clemensv) [reply http-protocol-binding.md:220]: "does not then need" dances appears to dance around least one of the ALLCAPS words. If SHOULD NOT is what you mean, that is what you should make it.

File: http-protocol-binding.md (hunk 7 above)
Comment (clemensv) [reply http-protocol-binding.md:225]: Better.

File: http-protocol-binding.md (hunk 8 above)
Comment (clemensv) [reply http-protocol-binding.md:228]: A detail of HTTP that I have not paid much attention to, indeed. I think that is a semantical issue for some HTTP field values rather than an encoding question, no?

File: http-protocol-binding.md (hunk 6 above)
Comment (jskeet) [reply http-protocol-binding.md:220]: This is what's in the new commit:

> The resulting string SHOULD NOT be further encoded.
> (Rationale: quoted string escaping is unnecessary when every space
> and double-quote character is already percent-encoded.)

File: http-protocol-binding.md (hunk 9 above)
Comment (jskeet) [reply http-protocol-binding.md:239]: Have added it speculatively.

File: http-protocol-binding.md (hunk 8 above)
Comment (jskeet) [reply http-protocol-binding.md:228]: It's part of RFC7230 section 3.2.6... so if we're telling SDK implementers they need to implement decoding as per that section of the RFC, it really helps if we're clear about what that means - in other words, that the comment part doesn't apply.

**Review by jskeet — COMMENTED** [review]
Pushing a new commit with comments addressed.

File: http-protocol-binding.md (hunk 9 above)
Comment (clemensv) [reply http-protocol-binding.md]: That's good. I would also further qualify this as "Unicode surrogate pair" in our text.

Comment (jskeet) [reply http-protocol-binding.md]: Done.

File: http-protocol-binding.md [hunk 10]
```diff
@@ -213,16 +213,75 @@ and per [RFC7230, section 3][rfc7230-section-3], HTTP headers MUST only use
...
+represented as HTTP header values MUST be percent-encoded as
+described below. This is compatible with [RFC3986, section
+2.1][rfc3986-section-2-1] but is more specific about what needs
+encoding. The resulting string SHOULD NOT be further encoded.
+(Rationale: quoted string escaping is unnecessary when every space
+and double-quote character is already percent-encoded.)
+
+When decoding an HTTP message into a CloudEvent, any HTTP header
+value MUST have quoted string escaping as described in [RFC7230,
+section 3.2.6][rfc7230-section-3-2-6] applied in reverse, and then a
```
Comment (duglin) [inline http-protocol-binding.md:226]: Might be ok, but the phrase "MUST have quoted string escaping ... applied in reverse" might be hard for some people to parse - it kind of sounds like we're saying it needs to be escaped but then it might not be clear what it means to say "in reverse".  So it's then odd since we can't have a MUST that applies to the sender if we're talking about the receiver/decoder. Yes I know we're just really trying to say they MUST de-escape - so perhaps we can rephrase this to be more clear?

Comment (jskeet) [reply http-protocol-binding.md]: I've had a go - see what you think. (I'm very happy to go round on this a few times. I don't want to leave any room for SDK implementers to be unclear about what they need to do.)

**Comment by duglin** [comment]
Approved on the 4/15 call

**Comment by jskeet** [comment]
Ah... this was going to go into master when everyone was happy with it. Because I suspect it should actually end up in 1.0.2.

I'll create a new PR for this on master, and we can decide whether to revert it from here.

**Comment by duglin** [comment]
oh crap - hold on

**Comment by duglin** [comment]
I reverted it on v1.0.1 via git

**Comment by duglin** [comment]
Sorry I mixed this one up with another PR and I didn't mean to merge it yet due to the edits made recently. But we have your other PR to use going forward.
